homeassistant/components/qnap_qsw/binary_sensor.py

```diff
@@ -0,0 +1,106 @@
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers.entity import EntityCategory
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from . import QswEntity
+from .const import ATTR_MESSAGE, DOMAIN
+from .coordinator import QswUpdateCoordinator
+
+
+@dataclass
+class QswBinarySensorEntityDescription(BinarySensorEntityDescription):
+    """A class that describes QNAP QSW binary sensor entities."""
+
+    attributes: dict[str, list[str]] | None = None
+    subkey: str = ""
+
+
+BINARY_SENSOR_TYPES: Final[tuple[QswBinarySensorEntityDescription, ...]] = (
+    QswBinarySensorEntityDescription(
+        attributes={
+            ATTR_MESSAGE: [QSD_FIRMWARE_CONDITION, QSD_MESSAGE],
+        },
+        device_class=BinarySensorDeviceClass.PROBLEM,
+        entity_category=EntityCategory.DIAGNOSTIC,
+        key=QSD_FIRMWARE_CONDITION,
+        name="Anomaly",
+        subkey=QSD_ANOMALY,
+    ),
+)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Add QNAP QSW binary sensors from a config_entry."""
+    coordinator: QswUpdateCoordinator = hass.data[DOMAIN][entry.entry_id]
+    async_add_entities(
+        QswBinarySensor(coordinator, description, entry)
+        for description in BINARY_SENSOR_TYPES
+        if (
+            description.key in coordinator.data
+            and description.subkey in coordinator.data[description.key]
+        )
+    )
+
+
+class QswBinarySensor(QswEntity, BinarySensorEntity):
+    """Define a QNAP QSW binary sensor."""
+
+    entity_description: QswBinarySensorEntityDescription
+
+    def __init__(
+        self,
+        coordinator: QswUpdateCoordinator,
+        description: QswBinarySensorEntityDescription,
+        entry: ConfigEntry,
+    ) -> None:
+        """Initialize."""
+        super().__init__(coordinator, entry)
+        self._attr_name = (
+            f"{self.get_device_value(QSD_SYSTEM_BOARD, QSD_PRODUCT)} {description.name}"
+        )
+        self._attr_unique_id = (
+            f"{entry.unique_id}_{description.key}_{description.subkey}"
+        )
+        self.entity_description = description
+        self._async_update_attrs()
+
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        """Update attributes when the coordinator updates."""
+        self._async_update_attrs()
+        super()._handle_coordinator_update()
+
+    @callback
+    def _async_update_attrs(self) -> None:
+        """Update binary sensor attributes."""
+        self._attr_is_on = self.get_device_value(
+            self.entity_description.key, self.entity_description.subkey
+        )
+
+        if self.entity_description.attributes:
+            self._attr_extra_state_attributes = {
+                key: self.get_device_value(val[0], val[1])
+                for key, val in self.entity_description.attributes.items()
+            }
```

Comment:
This code is the same as `sensor.py`. Can it be moved to a base class and call `super()._async_update_attrs`?

Reply:
Done in edd7a7f276c7886001acb6fef0e3b07d2109cf10